File: lib/contrek/finder/list.rb
module Contrek
  module Finder
    class List
      attr_reader :start, :end, :size, :idd
      def initialize(id)
        @start = nil
        @end = nil
        @size = 0
        @idd = id
      end

      def first
        @start
      end

      def find
        each { |e| return e if yield(e) }
        nil
      end

      def contains(entry)
        entry.data_pointer[@idd].inside
      end

      def grab(source_list)
        return if source_list.size == 0

        source_list_idd = source_list.idd
        source_list_start_entry = source_list.start

        act = source_list.end
        loop do
          break if (act = act.data_pointer[source_list_idd].prev).nil?
        end

        source_entry = source_list.start
        loop do
          source_entry.data_pointer[@idd] = source_entry.data_pointer[source_list_idd]
          next_entry = source_entry.data_pointer[source_list_idd].next
          source_entry.data_pointer[source_list_idd] = Contrek::Finder::Lists::Link.new(nil, nil, false)
          break if next_entry.nil?
          source_entry = next_entry
        end

        source_list_start_entry.data_pointer[@idd].prev = @end
        @end.data_pointer[@idd].next = source_list_start_entry unless @end.nil?

        @end = source_list.end
        @start = source_list.start if @start.nil?
        @size += source_list.size

        source_list.reset
      end

      def reset
        @end = nil
        @start = nil
        @size = 0
      end

      def <<(entry)
        return if entry.data_pointer[@idd].inside == true

        if @size > 0
          @end.data_pointer[@idd].next = entry
          entry.data_pointer[@idd].prev = @end
        else
          @start = entry
        end
        @end = entry
        entry.data_pointer[@idd].inside = true
        @size += 1
      end

      def shift
        return nil if @size == 0
        retme = @start
        next_of_retme = retme.data_pointer[@idd].next
        @start = next_of_retme

        @end = nil if retme == @end

        next_of_retme.data_pointer[@idd].prev = nil unless next_of_retme.nil?
        @size -= 1

        retme.data_pointer[@idd].next = nil
        retme.data_pointer[@idd].prev = nil
        retme.data_pointer[@idd].inside = false
        retme
      end

      def each
        return if @size == 0
        act = @start
        loop do
          yield act
          break if (act = act.data_pointer[@idd].next).nil?
        end
      end

      def map
        ary = []
        each { |e| ary << yield(e) }
        ary
      end

      def delete(entry)
        return if @size == 0
        return if entry.data_pointer[@idd].inside == false

        next_of_entry = entry.data_pointer[@idd].next
        prev_of_entry = entry.data_pointer[@idd].prev

        case entry
        when @start
          @start = next_of_entry
        when @end
          @end = prev_of_entry
        end

        next_of_entry.data_pointer[@idd].prev = prev_of_entry unless next_of_entry.nil?
        prev_of_entry.data_pointer[@idd].next = next_of_entry unless prev_of_entry.nil?

        entry.data_pointer[@idd].next = nil
        entry.data_pointer[@idd].prev = nil

        @size -= 1
        entry.data_pointer[@idd].inside = false
      end
    end
  end
end

File: ext/cpp_polygon_finder/PolygonFinder/src/polygon/finder/List.cpp
/*
 * List.cpp
 *
 * Copyright (c) 2025-2026 Emanuele Cesaroni
 *
 * Licensed under the GNU Affero General Public License v3 (AGPLv3).
 * See the LICENSE file in this directory for the full license text.
 */

#include "List.h"
#include <iostream>

List::List(int id) : start(nullptr), end(nullptr), isize(0), idd(id) {}

Listable *List::first() { return this->start; }
Listable *List::last() { return this->end; }
int List::get_id() { return this->idd; }
int List::size() { return this->isize; }

bool List::contains(Listable *entry) {
  return entry->data_pointer[this->idd].inside;
}

void List::reset() {
  this->end = nullptr;
  this->start = nullptr;
  this->isize = 0;
}

void List::push_back(Listable *entry) {
  auto& meta = entry->data_pointer[this->idd];
  if (meta.inside) return;

  if (this->isize > 0) {
    this->end->data_pointer[this->idd].next = entry;
    meta.prev = this->end;
  } else {
    this->start = entry;
    meta.prev = nullptr;
  }

  meta.next = nullptr;
  this->end = entry;
  meta.inside = true;
  this->isize++;
}

void List::remove(Listable *entry) {
  auto& meta = entry->data_pointer[this->idd];
  if (this->isize == 0 || !meta.inside) return;

  Listable *nxt = meta.next;
  Listable *prv = meta.prev;

  if (entry == this->start) this->start = nxt;
  if (entry == this->end)   this->end = prv;

  if (nxt) nxt->data_pointer[this->idd].prev = prv;
  if (prv) prv->data_pointer[this->idd].next = nxt;

  meta.next = nullptr;
  meta.prev = nullptr;
  meta.inside = false;
  this->isize--;
}

Listable *List::shift() {
  if (this->isize == 0) return nullptr;
  Listable *retme = this->start;
  auto& meta = retme->data_pointer[this->idd];

  this->start = meta.next;
  if (retme == this->end) this->end = nullptr;
  if (this->start) this->start->data_pointer[this->idd].prev = nullptr;

  this->isize--;
  meta.next = nullptr;
  meta.prev = nullptr;
  meta.inside = false;
  return retme;
}

void List::grab(List *source_list) {
  if (source_list->isize == 0) return;

  int src_idd = source_list->idd;
  Listable *current = source_list->start;

  while (current) {
    auto& src_meta = current->data_pointer[src_idd];
    auto& dst_meta = current->data_pointer[this->idd];
    dst_meta = src_meta;
    src_meta.inside = false;
    src_meta.next = nullptr;
    src_meta.prev = nullptr;
    current = dst_meta.next;
  }

  if (this->isize > 0) {
    this->end->data_pointer[this->idd].next = source_list->start;
    source_list->start->data_pointer[this->idd].prev = this->end;
    this->end = source_list->end;
  } else {
    this->start = source_list->start;
    this->end = source_list->end;
  }

  this->isize += source_list->isize;
  source_list->reset();
}
